docs/proposals/20210335-load-balancer-provider.md

```diff
@@ -0,0 +1,999 @@
+                    - requeue resource appropriately
+                - resource present and active in load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'active'
+            - `loadbalancer.cluster.x-k8s.io/disabled` is present:
+                - resource active in load balancer:
+                    - start removal/disabling of resource from lb
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabling'
+                    - requeue resource appropriately
+                - resource is being removed/disabled from lb:
+                    - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                    - requeue resource appropriately
+                - resource is removed/disabled from the load balancer:
+                    - set `loadbalancer.cluster.x-k8s.io/status` to 'disabled'
+        - resource is deleting:
+            - resource active in load balancer:
+                - start removal of resource from lb:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removing'
+                - requeue resource appropriately
+            - resource is being removed from lb:
+                - update `loadbalancer.cluster.x-k8s.io/status` appropriately for resource
+                - requeue resource appropriately
+            - resource is removed from load balancer:
+                - set `loadbalancer.cluster.x-k8s.io/status` to 'removed'
+
+- Ensure there are no resources configured in the load balancer that do not match the selector
+	
+
+LB provider registers pre-drain or pre-term hook on machines that are LoadBalancerMembers
+CP provider issues machine delete
+LB provider processes hook, quiesces, removes its hook annotation
+Machine is deleted
+
+
+### Example Provider (AWSELBClassic)
+In this example, we are provisioning a Kubernetes cluster on AWS to be backed by an Classic Elastic Load Balancer, which is represents the current behaviour of Cluster API Provider AWS.
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha4
+kind: Cluster
+metadata:
+  name: my-cluster
+  namespace: default
+spec:
+  clusterNetwork:
+    pods:
+      cidrBlocks:
+        - 192.168.0.0/16
+  controlPlaneRef:
+    apiGroup: controlplane.cluster.x-k8s.io
+    kind: KubeadmControlPlane
+    name: my-cluster
+  infrastructureRef:
+    apiGroup: infrastructure.cluster.x-k8s.io
+    kind: AWSCluster
+    name: my-cluster
+---
+apiVersion: controlplane.cluster.x-k8s.io/v1alpha4
+kind: KubeadmControlPlane
+metadata:
+  name: my-cluster
+  namespace: default
+spec:
+  infrastructureTemplate:
+    apiGroup: infrastructure.cluster.x-k8s.io
+    kind: AWSMachineTemplate
+    name: my-cluster
+  loadBalancerRef:
+    apiGroup: cluster.x-k8s.io
+    kind: LoadBalancer
+    name: my-cluster-control-plane-lb
+---
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: AWSCluster
+metadata:
+  name: my-cluster
+spec:
+  region: eu-west-1
+  sshKeyName: default
```

Comment:
Normally an AWSCluster would itself create a load balancer, how does it know not to do that in this case where it isn't needed?